File: easy/utopian_tree.py
from functools import reduce


def reducer(growth, year):
    if year % 2 == 0:
        return growth + 1
    else:
        return growth * 2


def utopian_tree(cycles):
    """
    DOCUMENTATION
    """
    growth = 1
    for cycle in range(1, cycles + 1):
        if cycle % 2 == 0:
            growth += 1
        else:
            growth *= 2
    return growth

File: src/easy/utopian-tree.ts
const spring = (value: number): number => value * 2;
const summer = (value: number): number => value + 1;
/**
 * @copyright https://www.hackerrank.com/challenges/utopian-tree/problem?utm_campaign=challenge-recommendation&utm_medium=email&utm_source=30-day-campaign
 * @param years
 * @returns
 */
export const utopianTree = (years: number): number => {
    return [...Array(years + 1).keys()].reduce(
        (growth: number, year: number) => {
            if (year === 0) return growth;
            if (year % 2 === 0) return summer(growth);
            return spring(growth);
        },
        1,
    );
};
